python/src/cloudstorage/cloudstorage_api.py

```diff
@@ -273,6 +277,193 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  (this will only need to happen once, since the file_list size restriction is 1024 = 32 * 32).
+  """
+  temp_list = []  # temporary storage for the filenames that store the merged segments of 32
+  if len(list_of_files) > 32:
+    temp_file_counter = 0
+    segments_list = [list_of_files[i:i + 32] for i in range(0, len(list_of_files), 32)]
+    if files_metadata is None:
+      files_metadata = []
+    meta_segments_list = [files_metadata[i:i + 32] for i in range(0, len(files_metadata), 32)]
+    files_metadata = []
+    list_of_files = []
+    for segment in segments_list:
+      if len(segment) > 1:
+        temp_file_name = destination_file + temp_file_suffix + str(temp_file_counter)
+        compose(segment, temp_file_name,
+                files_metadata=meta_segments_list,
+                content_type=content_type,
+                retry_params=retry_params,
+                _account_id=_account_id)
+        list_of_files.append(temp_file_name.replace(bucket, '', 1))
+        temp_file_counter += 1
+        temp_list.append(temp_file_name)
+      else:
+        list_of_files.append(segment[0])
+  # There will always be 32 or less files to merge at this point
+  compose(list_of_files, destination_file,
+          files_metadata=files_metadata,
+          content_type=content_type,
+          retry_params=retry_params,
+          _account_id=_account_id)
+  # Grab all temp files that were created during the merging of segments of 32
+  temp_list = listbucket(destination_file + temp_file_suffix)
+  # Delete all the now-unneeded temporary merge-files for the segments of 32 (if any)
+  for item in temp_list:
+    try:
+      delete(item.filename)
+    except errors.NotFoundError:
+      pass
+
+
+def _file_exists(destination):
+  """Checks if a file exists.
+  Tries to open the file.
+  If it succeeds returns True otherwise False.
+
+  Args:
+    destination: Full path to the file (ie. /bucket/object) with leading slash.
+
+  Returns:
+    True if the file is accessible otherwise False.
+  """
+  try:
+    with open(destination, "r") as _:
+      pass
+    return True
+  except errors.NotFoundError:
+    return False
+
+
+def _validate_compose_list(destination_file, file_list, files_metadata=None, number_of_files=32):
+  """Validates the compose list and builds and merges the file_list, files_metadata.
+
+  Args:
+    destination: Full path to the file (ie. /destination_bucket/destination_file).
+    file_list: List of files to compose, see compose for details.
+    files_metadata: Meta details for the file.
+    number_of_files: Maximum number of files allowed in the list.
+  Returns:
+    A tuple (list_of_files, bucket):
+      list_of_files: Ready to use dict version of the list.
+      bucket: bucket name extracted from the file paths.
+  """
+  common.validate_file_path(destination_file)
+  bucket = destination_file[0:(destination_file.index('/', 1) + 1)]
+  if not isinstance(file_list, list):
```

Comment:
unnecessary (will cause other "valid" types such as tuples, sets to be rejected)